cmd/pod-scaler/admission.go

```diff
@@ -87,15 +101,115 @@ func (m *podMutator) Handle(ctx context.Context, req admission.Request) admissio
 	return response
 }
 
-func mutatePod(pod *corev1.Pod, build *buildv1.Build) {
+// mutatePodMetadata updates metadata labels for Pods created by Prow for rehearsals,
+// where default metadata points to the release repo instead of the repo under test.
+// We can fix this by updating to use the values from the configuration that the job
+// ends up running with.
+func mutatePodMetadata(pod *corev1.Pod) error {
+	if _, isRehearsal := pod.ObjectMeta.Labels[rehearse.Label]; !isRehearsal {
+		return nil
+	}
+	var rawConfig string
+	for _, container := range pod.Spec.Containers {
+		if container.Name != "test" {
+			continue
+		}
+		for _, value := range container.Env {
+			if value.Name != "CONFIG_SPEC" {
+				continue
+			}
+			rawConfig = value.Value
+		}
+	}
+	if rawConfig == "" {
+		return errors.New("could not find configuration in rehearsal Pod's env")
+	}
+	var config api.ReleaseBuildConfiguration
+	if err := yaml.Unmarshal([]byte(rawConfig), &config); err != nil {
+		return fmt.Errorf("could not unmarshal configuration from rehearsal pod: %w", err)
+	}
+	pod.ObjectMeta.Labels[kube.ContextAnnotation] = pod.ObjectMeta.Labels[rehearse.LabelContext]
+	pod.ObjectMeta.Labels[kube.OrgLabel] = config.Metadata.Org
+	pod.ObjectMeta.Labels[kube.RepoLabel] = config.Metadata.Repo
+	pod.ObjectMeta.Labels[kube.BaseRefLabel] = config.Metadata.Branch
+	return nil
+}
+
+func mutatePodLabels(pod *corev1.Pod, build *buildv1.Build) {
 	if pod.Labels == nil {
 		pod.Labels = map[string]string{}
 	}
-	for _, label := range []string{steps.LabelMetadataOrg, steps.LabelMetadataRepo, steps.LabelMetadataBranch, steps.LabelMetadataVariant, steps.LabelMetadataTarget, steps.LabelMetadataStep} {
+	for _, label := range []string{steps.LabelMetadataOrg, steps.LabelMetadataRepo, steps.LabelMetadataBranch, steps.LabelMetadataVariant, steps.LabelMetadataTarget} {
 		buildValue, buildHas := build.Labels[label]
 		_, podHas := pod.Labels[label]
 		if buildHas && !podHas {
 			pod.Labels[label] = buildValue
 		}
 	}
 }
+
+// useOursIfLarger updates fields in theirs when ours are larger
+func useOursIfLarger(allOfOurs, allOfTheirs *corev1.ResourceRequirements) {
```

Comment:
I don't see how changing the semantics to return the max instead of mutating changes much here. I didn't want to have to copy-paste the logic 4 times (requests, limits, etc) which is why the abstraction over the ResourceList was written. Should I just copy-paste it?